benchmark/torch/RL4LMs/rl4lms_utils/rollout_util.py

```diff
@@ -0,0 +1,239 @@
+
+
+def unpack_observations(obs_tensor, n_instructors):
+    """
+    Unpacks vectorized dict observations into separate dict observations
+    """
+    unpacked_obs = []
+    keys = obs_tensor.keys()
+    for instructor_ix in range(n_instructors):
+        obs_dict = {}
+        for key in keys:
+            obs_dict[key] = obs_tensor[key][instructor_ix].reshape(1, -1).cpu()
+        unpacked_obs.append(obs_dict)
+    return unpacked_obs
+
+
+class RolloutUtil:
+    def __init__(self, kl_args):
+        self._kl_controller = KLController(kl_args["coeff"], kl_args["target_kl"])
+
+    def collect_rollouts(self, agent, instructor_group, rollout_buffer):
+        # get tokenizer
+        tokenizer = instructor_group.tokenizer
+
+        # Switch to eval mode both training and testing
+        agent.eval_mode()
+
+        # reset rollout buffer and stats
+        rollout_buffer.reset()
+
+        # start the rollout process
+        rollout_info = {
+            "rollout_info/ep_rew": [],
+            "rollout_info/kl_div_mean": [],
+            "rollout_info/ep_lens": [],
+            "rollout_info/ep_kl_rew": [],
+            "rollout_info/log_prob": [],
+            "rollout_info/ref_log_prob": [],
+            "rollout_info/values": [],
+        }
+        num_timesteps = 0
+        while not rollout_buffer.full:
+            # start parallel episodes
+            current_obs = instructor_group.ask()
+
+            # generate sentences using the model
+            generation_inputs = agent.get_inputs_for_generation(current_obs)
+
+            # note: RL4LMs uses the same way (language model always does sample() to generate in summarization
+            #       task) for collecting data and testing, so here agent uses predict() rather than sample()
+            gen_output = agent.predict(
+                input_ids=generation_inputs.inputs,
+                attention_mask=generation_inputs.attention_masks,
+                tokenizer=tokenizer)
+
+            # get episode state, reward, dones, infos from instructors
+            sentence_new_obs, sentence_rewards, sentence_dones, sentence_infos = instructor_group.feedback_sentense(
+                gen_output=gen_output)
+
+            # generate batch of rollouts and add to buffer
+            episode_wise_transitions, run_timesteps = self._generate_transition(
+                gen_sentence=gen_output,
+                init_obs=current_obs,
+                agent=agent,
+                n_instructors=instructor_group.n_instructors,
+                obs_space=instructor_group.observation_space,
+                sentence_new_obs=sentence_new_obs,
+                sentence_rewards=sentence_rewards,
+                sentence_dones=sentence_dones,
+                sentence_infos=sentence_infos,
+            )
+            num_timesteps += run_timesteps
+
+            # now we flush all episode wise info to the 1-D buffer
+            # log transition and add to buffer
```

Comment:
Disscussion: Can we replace line 101~137 as `rollout_info = rollout_buffer.add_transitions(eposode_wise_transitions)` ?